### PR TITLE
Bugfix: Fleet jump for valid mission escorts

### PR DESCRIPTION
**Pull request by samrocketman**
Summary
-------

Bug described on discord https://discord.com/channels/251118043411775489/536900466655887360/990559943860060180

Marcus said:

> when you do shift j so all ships jump at the same time, any ships that are stalking you like the three ravens for deep archaeology will hold up the jump cause I guess they technically count as part of the fleet?

This is considered a bug.

Fix details
------

NPC escorts should not be considered unless they're mission NPCs the player is actively escorting.  The fix will skip all escorts which are not player owned if they are not mission NPCs being escorted by the player fleet.

Save file
-----

https://discord.com/channels/251118043411775489/536900466655887360/990633732610338866

I edited the above save file to create this save file https://gist.github.com/samrocketman/b2d25586101cd7886b220388beaa3e45

Keep jumping.  Eventually the Raven will need to stop for fuel.  Notice your escort fleet will not wait for it and keep jumping.

Performance considerations
-------

This should be more performant since it skips invalid escorts escorting the player ship (for example archons and others).

Testing done
---------

This was successfully tested and verified through break points inspecting game memory that the intended behavior is fixed.

### CODE REVIEW

**Comment by samrocketman** [comment]
Fix player-owned escort checking

- Apply escort filter only if leader ship is player-owned.
- Follow loop standards similar to other code in `AI.cpp`.  `escort` was renamed to `ptr`.  `locked` was renamed to `escort`.

File: source/AI.cpp [hunk 1]
```diff
@@ -89,12 +89,16 @@ namespace {
...
+		bool shipIsYours = ship.IsYours();
+		for(const weak_ptr<Ship> &ptr : ship.GetEscorts())
 		{
-			shared_ptr<const Ship> locked = escort.lock();
-			if(locked && !locked->IsDisabled() && !locked->CanBeCarried()
-					&& locked->GetSystem() == ship.GetSystem()
-					&& locked->JumpFuel() && !locked->IsReadyToJump(true))
+			shared_ptr<const Ship> escort = ptr.lock();
+			// Skip escorts which are not player-owned and not escort mission NPCs.
+			if(!escort || (shipIsYours && !escort->IsYours() && !escort->GetPersonality().IsEscort()))
```
Comment (tehhowch) [inline source/AI.cpp:97]: You also need to check the hostility of NPCs. It is perfectly valid to assign `personality escort` to a hostile NPC.

**Review by tehhowch — CHANGES_REQUESTED** [review]

File: source/AI.cpp (hunk 1 above)
Comment (samrocketman) [reply source/AI.cpp:97]: As in check if NPC government is hostile against player government?

Comment (samrocketman) [reply source/AI.cpp]: @tehhowch see d397ff89db0ebacf6d198eef3c0f285ca8c119d5; I think I have addressed this feedback.

File: source/AI.cpp [hunk 2]
```diff
@@ -89,12 +89,17 @@ namespace {
...
+		const Government *gov = ship.GetGovernment();
+		for(const weak_ptr<Ship> &ptr : ship.GetEscorts())
 		{
-			shared_ptr<const Ship> locked = escort.lock();
-			if(locked && !locked->IsDisabled() && !locked->CanBeCarried()
-					&& locked->GetSystem() == ship.GetSystem()
-					&& locked->JumpFuel() && !locked->IsReadyToJump(true))
+			shared_ptr<const Ship> escort = ptr.lock();
+			// Skip escorts which are not player-owned and not escort mission NPCs.
+			if(!escort || gov->IsEnemy(escort->GetGovernment()) || (shipIsYours && !escort->IsYours() && !escort->GetPersonality().IsEscort()))
```
Comment (quyykk) [inline source/AI.cpp:98]:
```suggestion
			if(!escort || (shipIsYours && !escort->IsYours() && (!escort->GetPersonality().IsEscort() || gov->IsEnemy(escort->GetGovernment()))))
```

I believe it should be like this.

Comment (samrocketman) [reply source/AI.cpp:98]: My original change should apply to NPCs as well.  An NPC shouldn't consider a hostile escort an escort.

Comment (samrocketman) [reply source/AI.cpp:98]: Is it possible for NPCs to have an escort that is hostile?

Comment (quyykk) [reply source/AI.cpp:98]: I don't think it is.

Comment (samrocketman) [reply source/AI.cpp:98]: I mean that makes sense; usually hostile escorts _I think_ are the result of player human mistakes.  @quyykk then do you still want rhis change?  If so I'll be able to commit it later in about 10h.

**Comment by samrocketman** [comment]
@tehhowch do my changes address your feedback?

**Pull request merged**